src/redownload/web_parsing.py

```diff
@@ -0,0 +1,58 @@
+""" The module for HTML parsing related functions. """
+import urllib.request
+
+import bs4
+
+from . import exceptions
+
+
+def download_from_url(url: str) -> bs4.BeautifulSoup:
+    """Downloads an HTML page from a url and converts it to a BeautifulSoup object.
+
+    :param url: The URL to download HTML from, in a string.
+    :return: BeautifulSoup object extracted from the url.
+    """
+    request = urllib.request.urlopen(url)
+    html_doc = request.read()
+    html_soup = bs4.BeautifulSoup(html_doc, features="html.parser")
+    return html_soup
+
+
+def extract_links(page: bs4.BeautifulSoup, extensions: list = None) -> list:
+    """Extracts all the links with a file extension listed in the extensions param from a BeautifulSoup object and
+    returns a list of them. Raises an exception if there are no audio links in the object. If extensions is not
+    specified, returns all links on the page.
+
+    :param page: a BeautifulSoup object to extract audio links from.
+    :param extensions: OPTIONAL: a list of file extensions to filter for. If not specified, returns all links.
+    :return: a list of links to audio files ending in .flac or .mp3
+    """
+    all_links = []
+    correct_links = []
+    # Get all links on page.
+    for link in page.findAll("link"):
+        all_links.append(link.get("href"))
+    for link in page.findAll("a"):
+        all_links.append(link.get("href"))
+
+    # FIXME - Somehow there are None objects at the end of the list.
+    for link in all_links:
+        if isinstance(link, type(None)):
+            all_links.remove(link)
```

Comment:
This might not work. Editing a list you're iterating over is usually bad. 

try: `all_links = [link for link in all_links if link is not None]` or `all_links = filter(bool, all_links)` This second one checks if the link is truthy, not if it's non-None. Not sure if that is okay.

Reply:
It needs to check if it's a nonetype object because somehow a couple nonetype objects are being added to the list and causing AttributeError s below. I'll look into doing this a different way.

Reply:
Are you familiar with the idea of things being truthy and falsey? Most things (like strings) are truthy, but None is falsey. `all_links = filter(bool, all_links)` will filter out everything that's falsey, which should just be the Nones.

You can get into trouble when using a technique like this to filter out Nones from numbers though, since 0 is falsey and it will be filtered out with the Nones. I don't think this applies here though

Reply:
Resolving as I removed this code.